server/protocol.go

```diff
@@ -104,134 +52,164 @@ type Bytestring struct {
 	Bytes []byte
 }
 
-func parseAttestEK(ek *attest.EK) (RegistrationData, error) {
+func parseAttestEK(ek *attest.EK) (EnrollData, error) {
 	if reflect.TypeOf(ek.Public).String() != "*rsa.PublicKey" {
-		return RegistrationData{}, errors.New("Invalid key type:" + reflect.TypeOf(ek.Public).String())
+		return EnrollData{}, errors.New("Invalid key type:" + reflect.TypeOf(ek.Public).String())
 	}
 	var c []byte = make([]byte, 0)
 	if ek.Certificate != nil {
 		c = ek.Certificate.Raw
 	}
 	var n = ek.Public.(*rsa.PublicKey).N.Bytes()
 	var e = ek.Public.(*rsa.PublicKey).E
-	return RegistrationData{c, n, e, *pcrextend}, nil
+	return EnrollData{c, n, e, *pcrextend}, nil
+}
+
+func establishEncryptedSession(ch chan []byte) (*Session, error) {
+	var data Bytestring
+	var key []byte
+	var session = NewSession(ch)
+	var err error
+
+	logrus.Info("Getting client UUID")
+	if err = recvMsg(&data, session); err != nil {
+		return nil, err
+	}
+	if session.uuid, err = uuid.FromBytes(data.Bytes); err != nil {
+		close(ch)
+		return nil, err
+	}
+
+	if *enroll {
+		// TODO: Save key somewhere
+	} else {
+		// TODO: Fetch key from persistence point
+	}
+	if err != nil {
+		close(ch)
+		return nil, err
+	}
+	if session.Encrypt(key); err != nil {
+		close(ch)
+		return nil, err
+	}
+	return session, nil
 }
 
-func sendRegistrationData(ch chan []byte, tpm *attest.TPM) error {
+func enrollment(session *Session, tpm *attest.TPM) error {
 	logrus.Info("Retrieving EK pub and EK cert")
 	eks, err := tpm.EKs()
 	if err != nil {
-		close(ch)
+		close(session.ch)
 		return err
 	}
-	logrus.Info("Sending EK pub and EK cert")
+	logrus.Info("Sending enrollment data")
 
 	// Any key should do the job in principle, use the first one
 	// as we expect it to include a certificate.
 	ek, err := parseAttestEK(&eks[0])
 	if err != nil {
-		close(ch)
+		close(session.ch)
 		return nil
 	}
-	err = sendMsg(ek, ch)
+	err = sendMsg(ek, session)
 	if err != nil {
 		return err
 	}
 	return nil
 }
 
-func authentication(ch chan []byte) error {
-	logrus.Info("Starting authentication process")
-	logrus.Info("Generating nonce")
-	b, err := getTPMRandom(16)
+func authentication(session *Session) error {
```

Comment:
This code was originally implemented to prevent a malicious user from performing slow cryptographic operations such as tpm2_createak(). I don't know if we gain in performance in view of the TPM operations performed by this countermeasure.

However this sort of authentication adds an early anti-replay layer. We could keep it afterthought.

Reply:
We finally decided to remove the authentication message.